apps/service_providers/tests/test_assistant_runnable.py

```diff
@@ -211,10 +234,112 @@ def test_assistant_uploads_new_file(create_and_run, retrieve_run, list_messages,
+    assert "openai-file-2" in message.metadata["openai_file_ids"]
+
+
+@pytest.mark.django_db()
+@patch("apps.service_providers.llm_service.runnables.AssistantExperimentRunnable._get_response_with_retries")
+@patch("openai.resources.beta.threads.runs.Runs.retrieve")
+@patch("openai.resources.beta.Threads.create_and_run")
+@patch("openai.resources.files.Files.retrieve")
+@patch("openai.resources.files.Files.retrieve_content")
+@patch("openai.resources.beta.threads.messages.Messages.list")
+def test_assistant_reponse_with_annotations(
+    list_messages, retrieve_file_content, retrieve_file, create_and_run, retrieve_run, get_response_with_retries
+):
+    """Test that attachments on AI messages are being saved (only those of type `file_path`)
+    OpenAI doesn't allow you to fetch the content of the file that you uploaded, but this isn't an issue, since we
+    already have that file as an attachment on the chat object
+    """
+    from langchain.agents.openai_assistant.base import OpenAIAssistantFinish
+    from openai.types.file_object import FileObject
+
+    file_path_file = FileObject(
+        id="openai-file-1",
+        bytes=69,
+        created_at=123,
+        filename="generated.txt",
+        object="file",
+        purpose="assistants",
+        status="processed",
+        status_details=None,
+    )
+    file_citation_file = FileObject(
+        id="openai-file-2",
+        bytes=69,
+        created_at=123,
+        filename="existing.txt",
+        object="file",
+        purpose="assistants",
+        status="processed",
+        status_details=None,
+    )
+    retrieve_file.side_effect = [file_path_file, file_citation_file]
+
+    retrieve_file_content.return_value = "This is some file content"
+    thread_id = "test_thread_id"
+    run = _create_run(ASSISTANT_ID, thread_id)
+    get_response_with_retries.return_value = OpenAIAssistantFinish(
+        run_id=run.id, thread_id=thread_id, return_values={"output": "Hi there human"}, log=""
+    )
+    existing_file = FileFactory(external_id="openai-file-2")
+    annotations = [
+        FilePathAnnotation(
+            end_index=174,
+            file_path=FilePath(file_id="openai-file-1"),
+            start_index=134,
+            text="sandbox:/mnt/data/random_string_file.txt",
+            type="file_path",
+        ),
+        FileCitationAnnotation(
+            end_index=174,
+            file_citation=FileCitation(file_id=existing_file.external_id, quote=""),
+            start_index=134,
+            text="[some-ref]",
+            type="file_citation",
+        ),
+    ]
+    list_messages.return_value = SyncCursorPage(
+        data=[
+            ThreadMessage(
+                id="test",
+                assistant_id=ASSISTANT_ID,
+                metadata={},
+                created_at=0,
+                content=[TextContentBlock(text=Text(annotations=annotations, value="Hi there human"), type="text")],
```

Comment:
the message text here should contain the citation references to test that they get replaced correctly